### PR TITLE
Add .sizeIs / .isEmpty / .head operators including NonEmpty types and improve type inference for constants

### PR DESCRIPTION
**Pull request by jeffmay**

### CODE REVIEW

File: circe-v1/src/test/scala/circe/StandardResultAsJsonSpec.scala [hunk 1]
```diff
@@ -14,14 +14,14 @@ class StandardResultAsJsonSpec extends FunSuite {
   import dsl.circe.standard._
 
   test("serialize an expression state without input") {
-    val expr = "World".const
+    val expr = const("World")(constId).const
```
Comment (jeffmay) [inline circe-v1/src/test/scala/circe/StandardResultAsJsonSpec.scala]: BLKR: Revert

File: core-v1/src/main/scala/algebra/SizeComparable.scala [hunk 2]
```diff
@@ -0,0 +1,73 @@
+package com.rallyhealth.vapors.v1
+
+package algebra
+
+import cats.data.{NonEmptyList, NonEmptySeq, NonEmptyVector}
+
+trait SizeComparable[-C, -N, +B] { outer =>
```
Comment (jeffmay) [inline core-v1/src/main/scala/algebra/SizeComparable.scala:15]: BLKR: Document. Also, explain why contravariant.

File: core-v1/src/main/scala/algebra/Expr.scala [hunk 3]
```diff
@@ -966,6 +978,18 @@ object Expr {
       copy(debugging = debugging)
   }
 
+  final case class SizeIs[-I, N : ExtractValue.As[Int]#From, B : ExtractValue.As[Boolean]#From : OP, OP[_]](
```
Comment (jeffmay) [inline core-v1/src/main/scala/algebra/Expr.scala]: BLKR: Add documentation

File: core-v1/src/main/scala/data/ExtractValue.scala [hunk 4]
```diff
@@ -15,9 +15,17 @@ trait ExtractValue[T, V] extends (T => V) {
 object ExtractValue {
   type AsBoolean[T] = ExtractValue[T, Boolean]
 
+  type As[V] = {
+    type From[T] = ExtractValue[T, V]
+  }
+
+  type From[T] = {
+    type As[V] = ExtractValue[T, V]
+  }
```
Comment (jeffmay) [inline core-v1/src/main/scala/data/ExtractValue.scala]: Are these being used? If so, I should document them and maybe refactor old code to use them where appropriate. I suppose one advantage of these is that you don't need the kind-projector compiler plugin to add these as context bounds. However, long-term, I think these will be phased out in favor of kind-projector... so why add them now? Are they really used enough to make it worth adding them here?

File: core-v1/src/main/scala/dsl/ConstOutputType.scala [hunk 5]
```diff
@@ -15,16 +15,20 @@ import scala.annotation.implicitNotFound
...
+)
+trait ConstOutputType[W[+_], -O] {
   type Out
 
   def wrapConst(value: O): Out
 }
 
 object ConstOutputType {
-  type Aux[W[+_], O, WO] = ConstOutputType[W, O] { type Out = WO }
+  type Aux[W[+_], -O, WO] = ConstOutputType[W, O] { type Out = WO }
```
Comment (jeffmay) [inline core-v1/src/main/scala/dsl/ConstOutputType.scala:33]: I found this to be helpful when determining the return type of the `.const` operator when working with `Some[A]`, as an implicit for `Option[A]` will be more specific (according to the rules of contravariance) than `Some[A]`, so it allow (and prefer) a `Traverse[Option]` over a `Traverse[Some]` (which does not exist)

Comment (jeffmay) [reply core-v1/src/main/scala/dsl/ConstOutputType.scala:33]: Also, I should move this comment about variance to the ScalaDocs

**Review by vinit-garg — APPROVED** [review]

**Review by mlinenweber — APPROVED** [review]

**Pull request merged**